### PR TITLE
wip: stream prototype

### PR DESCRIPTION
**Pull request by eddort**
WIP: stream prototype
TODO:
- prepare RFC on methods of working with streams
- scale the current solution to all endpoints

### CODE REVIEW

File: src/common/registry/storage/key.storage.ts [hunk 1]
```diff
@@ -16,6 +16,27 @@ export class RegistryKeyStorageService {
     return await this.repository.find(where, options);
   }
 
+  async *fetchKeysByChunks(where: FilterQuery<RegistryKey>, options: any): AsyncGenerator<any, void, unknown> {
+    const batchSize = 10000;
+    let offset = 0;
+    // TODO: transaction
+    while (true) {
+      const query = this.repository.createQueryBuilder().select('*').where(where).limit(batchSize).offset(offset);
```
Comment (sergeyWh1te) [inline src/common/registry/storage/key.storage.ts:24]: * - most cost operation for databases

File: src/common/registry/storage/key.storage.ts [hunk 2]
```diff
@@ -16,6 +16,27 @@ export class RegistryKeyStorageService {
...
+
+      const chunk = await query.execute();
+
+      if (chunk.length === 0) {
+        break;
+      }
+
+      offset += batchSize;
```
Comment (sergeyWh1te) [inline src/common/registry/storage/key.storage.ts:32]: For better performance in db. Better and easy to use cursors instead limit&offset approach

Example:
https://www.geeksforgeeks.org/postgresql-cursor/

File: src/common/registry/storage/key.storage.ts (hunk 1 above)
Comment (eddort) [reply src/common/registry/storage/key.storage.ts:24]: we will check after the merge, thanks!

File: src/common/registry/storage/key.storage.ts (hunk 2 above)
Comment (eddort) [reply src/common/registry/storage/key.storage.ts:32]: MikroORM doesn't support it, I wrote it in the RFC

**Comment by eddort** [comment]
Trying to re-implement it with another lib
```js
const StreamArray = require('stream-json/streamers/StreamArray');
const {chain}  = require('stream-chain');
const {parser} = require('stream-json');

// Assume that keysStream is a stream returning key objects
let pipeline = chain([
  keysStream,
  parser(),
  new StreamArray(),
]);

// Create a function that yields a JSON string with meta information and keys
async function* generateJSON() {
  yield '{ "meta": ' + JSON.stringify(meta) + ', "keys": [';
  
  for await (const keysBatch of pipeline) {
    yield JSON.stringify(keysBatch.value) + ',';
  }

  yield ']}';
}

// Send the stream as a response
reply.type('application/json').send(generateJSON());

```

- Since we need to collect json of complex shapes we have to write an additional function on generators, and the use of this library is reduced to the fact that we organize more beautiful chain
- It is possible not to use libraries at all, but to write your own generator.

**Pull request merged**